discovery-provider/es-indexer/src/conn.ts

```diff
@@ -0,0 +1,84 @@
+import { Pool as PG } from 'pg'
+import { Client as ES } from '@elastic/elasticsearch'
+import Cursor from 'pg-cursor'
+import _ from 'lodash'
+import { indexNames } from './indexNames'
+import { BlocknumberCheckpoint } from './types/blocknumber_checkpoint'
+
+let pool: PG | undefined = undefined
+export function dialPg(): PG {
+  if (!pool) {
+    let connectionString = process.env.audius_db_url
+    pool = new PG({ connectionString, application_name: 'es-indexer' })
+  }
+
+  return pool
+}
+
+export async function queryCursor(sql: string) {
+  const client = await dialPg().connect()
+  const cursor = client.query(new Cursor(sql))
+  return { client, cursor }
+}
+
+let esc: ES | undefined = undefined
+export function dialEs() {
+  if (!esc) {
+    let url = process.env.audius_elasticsearch_url
+    esc = new ES({ node: url })
+  }
+  return esc
+}
+
+export async function waitForHealthyCluster() {
+  return dialEs().cluster.health(
+    {
+      wait_for_status: 'green',
+      timeout: '60s',
+    },
+    {
+      requestTimeout: '60s',
+    }
+  )
+}
+
+/**
+ * Gets the max(blocknumber) from elasticsearch indexes
+ * Used for incremental indexing to understand "where we were" so we can load new data from postgres
+ */
+export async function getBlocknumberCheckpoints(): Promise<BlocknumberCheckpoint> {
+  const searches = []
+
+  for (let [tableName, indexName] of Object.entries(indexNames)) {
+    searches.push({ index: indexName })
+    searches.push({
+      size: 0,
+      aggs: {
+        max_blocknumber: {
+          max: {
+            field: checkpointField(tableName),
+          },
+        },
+      },
+    })
+  }
+
+  const multi = await dialEs().msearch({ searches })
+
+  const values = multi.responses.map(
+    (r: any) => r.aggregations?.max_blocknumber?.value || 0
```

Comment:
can we type this?

Reply:
it is possible, but forces you to discriminate type union for handling the error response... which is good, but for me this falls below the cost/benefit considering the `?` operator effectively gives you the desired behavior.

Reply:
well elasticsearch does have decent typing, but it can't read into your query to determine what fields will be present on the result... for instance `max_blocknumber` is a agg name that I typed.

Here was my attempt to type... but the possibility of the error response in the union was when I decided I didn't care anymore haha:

<img width="867" alt="image" src="https://user-images.githubusercontent.com/97163/165601975-b76a2c23-38be-48c9-9f6c-ebf73caa7f23.png">